firestore/src/jni/arena_ref.cc

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright 2023 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "firestore/src/jni/arena_ref.h"
+
+#include <atomic>
+#include <utility>
+
+#include "firestore/src/android/firestore_android.h"
+#include "firestore/src/jni/env.h"
+#include "firestore/src/jni/hash_map.h"
+#include "firestore/src/jni/long.h"
+#include "firestore/src/jni/map.h"
+
+namespace firebase {
+namespace firestore {
+namespace jni {
+
+namespace {
+
+HashMap* gArenaRefHashMap = nullptr;
+jclass gHashMapClass = nullptr;
+jclass gLongClass = nullptr;
+jmethodID gLongConstructor = nullptr;
+std::mutex mutex_;
+
+int64_t GetNextArenaRefKey() {
+  static std::atomic<int64_t> next_key(0);
+  return next_key.fetch_add(1);
+}
+
+}  // namespace
+
+ArenaRef::ArenaRef(Env& env, const Object& object)
+    : key_(GetNextArenaRefKey()) {
+  std::unique_lock<std::mutex> lock(mutex_);
+  gArenaRefHashMap->Put(env, key_object(env), object);
+}
+
+ArenaRef::ArenaRef(const ArenaRef& other)
+    : key_(other.key_ == -1 ? -1 : GetNextArenaRefKey()) {
+  if (other.key_ != -1) {
+    Env env = FirestoreInternal::GetEnv();
+    Local<Object> object = other.get(env);
+    std::unique_lock<std::mutex> lock(mutex_);
+    gArenaRefHashMap->Put(env, key_object(env), object);
+  }
+}
+
+ArenaRef::ArenaRef(ArenaRef&& other) { std::swap(key_, other.key_); }
```

Comment:
Please change this to:

```cpp
using std::swap;  // go/using-std-swap
swap(key_, other.key_);
```

I just stumbled across this advice. It probably doesn't technically matter here since we're swapping ints, but it's good to showcase best practices.